skbio/stats/composition.py

```diff
@@ -2058,3 +2061,423 @@ def dirmult_ttest(
+    ...         "x3": [1.09999862, 2.99998318],
+    ...         "y1": [1.00000084, 2.10001257],
+    ...         "y2": [0.9999991, 3.09998418],
+    ...         "y3": [0.99999899, 3.9999742],
+    ...         "z1": [1.10000124, 5.0001796],
+    ...         "z2": [1.00000053, 6.09924644],
+    ...         "z3": [1.10000173, 6.99693644],
+    ...     },
+    ...     index=["Y1", "Y2"],
+    ... ).T
+    >>> metadata = pd.DataFrame(
+    ...     {
+    ...         "patient": [1, 1, 1, 2, 2, 2, 3, 3, 3, 4, 4, 4],
+    ...         "treatment": [1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2],
+    ...         "time": [1, 2, 3, 1, 2, 3, 1, 2, 3, 1, 2, 3],
+    ...     },
+    ...     index=["x1", "x2", "x3", "y1", "y2", "y3",
+    ...            "z1", "z2", "z3", "u1", "u2", "u3"],
+    ... )
+    >>> res = dirmult_lme(
+    ...     formula="time + treatment", data=table, metadata=metadata,
+    ...     groups="patient", seed=0, p_adjust="sidak"
+    ... )
+    >>> res
+      FeatureID  Covariate  Log2(FC)   CI(2.5)  CI(97.5)    pvalue    qvalue
+    0        Y1       time -0.210769 -1.532142  1.122417  0.403737  0.873598
+    1        Y1  treatment -0.744093 -3.401875  1.582636  0.252057  0.687051
+    2        Y2       time  0.210769 -1.122417  1.532142  0.403737  0.873598
+    3        Y2  treatment  0.744093 -1.582636  3.401875  0.252057  0.687051
+
+    """
+
+    # Test if data is a pandas DataFrame
+    errmsg = (
+        "%s must be a pandas DataFrame or a numpy structured or rec array or "
+        "a dictionary."
+    )
+
+    if not isinstance(data, pd.DataFrame):
+        try:
+            data = pd.DataFrame(data)
+        except (TypeError, ValueError):
+            raise TypeError(errmsg % "Data")
+
+    if data.ndim != 2:
+        try:
+            data = pd.DataFrame(data, index=data.dtype.names)
+        except (TypeError, ValueError):
+            raise TypeError(errmsg % "Data")
+
+    if not hasattr(data, "__getitem__"):
+        try:
+            data = pd.DataFrame(data)
+        except (TypeError, ValueError):
+            raise TypeError(errmsg % "Data")
+
+    if not isinstance(metadata, pd.DataFrame):
+        try:
+            metadata = pd.DataFrame(metadata)
+        except (TypeError, ValueError):
+            raise TypeError(errmsg % "Metadata")
+
+    # Test if data has missing values
+    if data.isnull().values.any():
+        raise ValueError("Cannot handle missing values in data.")
+
+    # Test if metadata has missing values
+    if metadata.isnull().values.any():
+        raise ValueError("Cannot handle missing values in metadata.")
+
+    # Test if metadata and data have the same index, regardless of order
+    if not data.index.sort_values().equals(metadata.index.sort_values()):
+        raise ValueError("Data and metadata must have the same index.")
+
+    rng = get_rng(seed)
+
+    posterior = [
+        rng.dirichlet(data.values[i] + pseudocount) for i in range(data.shape[0])
+    ]
```

Comment:
I think that should be equivalent too the below. But, note that the operation may be more efficient if the frame is expressed columnar as that is how Pandas stores data

```suggestion
    def dirichlet_with_pseudocount(pseudocount, row):
        return rng.dirichlet(row + pseudocount)
    dirichlet_f = functools.partial(dirichlet_with_pseudocount, pseudocount)
    posterior = data.apply(dirichlet_f, axis=1) 
```